modules/api/php/login.class.inc

```diff
@@ -0,0 +1,207 @@
+            } else {
+                return (new \LORIS\Http\Response())
+                    ->withBody(
+                        new \LORIS\Http\StringStream(
+                            ['error' => 'Unacceptable JWT key']
+                        )
+                    )
+                    ->withHeader("Content-Type", "application/json")
+                    ->withStatus(500);
+            }
+        }
+        return (new \LORIS\Http\Response())
+            ->withBody(
+                new \LORIS\Http\StringStream(
+                    json_encode(array("error" => $login->_lastError))
+                )
+            )
+            ->withHeader("Content-Type", "application/json")
+            ->withStatus(401);
+    }
+
+    /**
+     * Get the SinglePointLogin class used to authenticate this request
+     * in a separate function so that it can be mocked out for testing.
+     *
+     * @return \SinglePointLogin Loris Authenticator
+     */
+    function getLoginAuthenticator()
+    {
+        return new \SinglePointLogin();
+    }
+
+    /**
+     * Return a valid JWT encoded identification token for the user
+     *
+     * @param string $user The user to return an identification token for
+     *
+     * @return string JWT encoded token
+     */
+    function getEncodedToken($user)
+    {
+        $factory = \NDB_Factory::singleton();
+        $config  = $factory->config();
+
+        $www     = $config->getSetting("www");
+        $baseURL = $www['url'];
+
+        $token = array(
+            // JWT related tokens to for the JWT library to validate
+                  "iss"  => $baseURL,
+                  "aud"  => $baseURL,
+            // Issued at
+                  "iat"  => time(),
+                  "nbf"  => time(),
+            // Expire in 1 day
+                  "exp"  => time() + 86400,
+            // Additional payload data
+                  "user" => $user,
+                 );
+
+        $key = $config->getSetting("JWTKey");
+        if (!self::isKeyStrong($key)) {
+            return "";
+        }
+        return \Firebase\JWT\JWT::encode($token, $key, "HS256");
+    }
+
+    /**
+     * Verify key meets cryptographic strength requirements
+     *
+     * @param string $key The JWT key to verify
+     *
+     * @return boolean Key passes strength test
+     */
+    static function isKeyStrong($key)
+    {
+        // Note: this code adapted from User::isPasswordStrong
+        $CharTypes = 0;
+        // less than 20 characters
+        if (strlen($key) < 20) {
+            return false;
+        }
+        // nothing but letters
+        if (!preg_match('/[^A-Za-z]/', $key)) {
+            return false;
+        }
+        // nothing but numbers
+        if (!preg_match('/[^0-9]/', $key)) {
+            return false;
+        }
+        // preg_match returns 1 on match, 0 on non-match
+        $CharTypes += preg_match('/[0-9]+/', $key);
+        $CharTypes += preg_match('/[A-Za-z]+/', $key);
```

Comment:
Could be split to force lower AND upper characters

Reply:
This is directly copied from the existing function in `htdocs/api/v0.0.2/Login.php`, just moved to a new home.